src/java/org/apache/cassandra/index/sai/disk/PrimaryKeyMapIterator.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.disk;
+
+import java.io.IOException;
+
+import org.apache.cassandra.db.PartitionPosition;
+import org.apache.cassandra.dht.AbstractBounds;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.index.sai.SSTableContext;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.index.sai.utils.RangeIterator;
+
+/**
+ * Iterates keys in the PrimaryKeyMap of an SSTable.
+ * Iterating keys in the primary key map is faster than reading them from the SSTable data component
+ * because we don't deserialize all the other columns except the primary key columns.
+ * The primary key map components are also likely much smaller than the whole SSTable data component.
+ * <p>
+ * The keys are returned in token-clustering order.
+ */
+public final class PrimaryKeyMapIterator extends RangeIterator
+{
+    private final PrimaryKeyMap keys;
+    private long currentRowId;
+
+    private PrimaryKeyMapIterator(PrimaryKeyMap keys, PrimaryKey min, PrimaryKey max)
+    {
+        super(min, max, keys.count());
+        this.keys = keys;
+        this.currentRowId = keys.exactRowIdForPrimaryKey(min);
```

Comment:
First, I think this lookup can be expensive, so if there is a way to pass the row id in as an argument, we should do that. Second, I think we need to handle the case where `min` does not have a row id because the current behavior is to return -1, and that'll probably lead to issues based on the definition of `computeNext`.

Reply:
Indeed, I pushed a change. Is it better now?